File: src/main/java/itmo/is/project/model/module/storage/StorageModuleFreeSpace.java
package itmo.is.project.model.module.storage;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.util.Pair;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class StorageModuleFreeSpace {

    @NotNull
    private StorageModule storageModule;

    @Min(0)
    @NotNull
    private Integer freeSpace;

    public StorageModuleFreeSpace(Pair<StorageModule, Integer> storageModuleFreeSpacePair) {
        this.storageModule = storageModuleFreeSpacePair.getFirst();
        this.freeSpace = storageModuleFreeSpacePair.getSecond();
    }

    public Integer getStorageModuleId() {
        return storageModule.getId();
    }

    public void add(int space) {
        freeSpace += space;
    }

    public void sub(int space) {
        freeSpace -= space;
    }
}

File: src/main/kotlin/itmo/isproject/storage/model/StorageModuleFreeSpace.kt
package itmo.isproject.storage.model

import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull

data class StorageModuleFreeSpace(
    var storageModule: @NotNull StorageModule? = null,
    var freeSpace: @Min(0) @NotNull Int = 0
) {

    constructor(storageModuleFreeSpacePair: Pair<StorageModule, Int>) : this(
        storageModule = storageModuleFreeSpacePair.first,
        freeSpace = storageModuleFreeSpacePair.second
    )

    operator fun plusAssign(amount: Int) {
        this.freeSpace = this.freeSpace.plus(amount)
    }

    operator fun minusAssign(amount: Int) {
        this.freeSpace = this.freeSpace.minus(amount)
    }
}
